File: plugins/shipyard/project-files/scripts/on-commit.py
#!/usr/bin/env python3
"""Post-commit hook: reset loop detection and trigger learning capture.

Called after a successful git commit to:
1. Reset the loop detection counter for committed files
2. Detect if a struggle just resolved (files that hit the loop threshold)
3. Signal the agent to capture what it learned into .claude/rules/learnings/

STDOUT CONTRACT: PostToolUse hooks send stdout as conversation messages to Claude.
Learning capture prompts go to stdout intentionally. Diagnostics/errors go to stderr.
"""

import json
import os
import re
import subprocess
import sys
import tempfile

SHIPYARD_DATA = os.environ.get('SHIPYARD_DATA', '.shipyard')
LOOP_STATE = os.path.join(SHIPYARD_DATA, '.loop-state.json')
LEARNINGS_DIR = '.claude/rules/learnings'


def sanitize_for_claude(s, max_len=200):
    """Strip control chars and cap length before printing untrusted strings
    to hook stdout. Hook stdout enters Claude's context, so file paths and
    git output (attacker-influenceable) need sanitization to prevent
    indirect prompt injection.
    """
    if not isinstance(s, str):
        s = str(s)
    s = ''.join(c for c in s if c == ' ' or (c.isprintable() and c != '\x7f'))
    if len(s) > max_len:
        s = s[:max_len] + '…[truncated]'
    return s


def reset_loop_state():
    """Reset edit counters and detect resolved struggles."""
    if not os.path.exists(LOOP_STATE):
        return []

    # Get files in the last commit
    try:
        result = subprocess.run(
            ['git', 'diff-tree', '--no-commit-id', '--name-only', '-r', 'HEAD'],
            capture_output=True, text=True
        )
    except FileNotFoundError:
        print("⚠️  on-commit: git not found on PATH", file=sys.stderr)
        return []
    if result.returncode != 0:
        print(f"⚠️  on-commit: git diff-tree failed: {result.stderr.strip()}", file=sys.stderr)
        return []
    committed_files = set(f for f in result.stdout.strip().split('\n') if f.strip())

    try:
        with open(LOOP_STATE) as f:
            state = json.load(f)
    except (json.JSONDecodeError, ValueError) as e:
        print(f"⚠️  on-commit: Corrupt loop state, skipping reset: {e}", file=sys.stderr)
        return []
    except OSError as e:
        print(f"⚠️  on-commit: Cannot read loop state: {e}", file=sys.stderr)
        return []

    edits = state.get('edits', {})
    struggles = state.get('struggles', {})

    # Find struggles that just resolved (committed files that were struggling)
    resolved_struggles = []
    for file_path in committed_files:
        if file_path in struggles:
            resolved_struggles.append({
                'file': file_path,
                'edit_count': struggles[file_path].get('edit_count', 0)
            })

    # Reset counters for committed files
    for f in committed_files:
        edits.pop(f, None)
        struggles.pop(f, None)

    state['edits'] = edits
    state['struggles'] = struggles

    # Atomic write to prevent corruption
    dir_path = os.path.dirname(LOOP_STATE)
    os.makedirs(dir_path, exist_ok=True)
    try:
        fd, tmp_path = tempfile.mkstemp(dir=dir_path, suffix='.json')
        with os.fdopen(fd, 'w') as f_out:
            json.dump(state, f_out, indent=2)
        os.replace(tmp_path, LOOP_STATE)
    except OSError as e:
        print(f"⚠️  on-commit: Failed to save loop state: {e}", file=sys.stderr)

    return resolved_struggles


def detect_domain(file_path):
    """Suggest a learnings category based on the file path."""
    path_lower = file_path.lower()

    # Map file paths to learning domains
    domain_patterns = [
        (['auth', 'login', 'session', 'token', 'jwt', 'oauth'], 'auth'),
        (['api', 'route', 'endpoint', 'handler', 'controller'], 'api'),
        (['test', 'spec', '__test', '.test.', '.spec.'], 'testing'),
        (['component', 'page', 'layout', 'view', 'ui'], 'ui'),
        (['style', 'css', 'scss', 'tailwind', 'theme'], 'styling'),
        (['model', 'schema', 'migration', 'database', 'db', 'query', 'supabase', 'prisma'], 'data'),
        (['hook', 'context', 'provider', 'store', 'state', 'redux', 'zustand'], 'state'),
        (['config', 'env', '.config', 'setting'], 'config'),
        (['util', 'helper', 'lib', 'service', 'action'], 'logic'),
    ]

    for keywords, domain in domain_patterns:
        if any(kw in path_lower for kw in keywords):
            return domain

    return 'general'


def signal_learning_capture(resolved):
    """Print a message prompting the agent to capture learnings."""
    if not resolved:
        return

    files = [s['file'] for s in resolved]
    max_edits = max(s['edit_count'] for s in resolved)

    # Suggest domains based on file paths
    domains = set(detect_domain(f) for f in files)
    domain_hint = ', '.join(sorted(domains))

    # Sanitize file paths before echoing to Claude (prompt injection defense)
    safe_files = [sanitize_for_claude(f) for f in files]

    print("")
    print("📝 LEARNING OPPORTUNITY — You just resolved a struggle.")
    print(f"   Files: {', '.join(safe_files)}")
    print(f"   Edits before resolution: {max_edits}")
    print(f"   Suggested domain(s): {domain_hint}")
    print("")
    print(f"   Append what you learned to .claude/rules/learnings/<domain>.md")
    print(f"   (create the file if it doesn't exist)")
    print("")
    print("   Format each entry as:")
    print("   ### [Short title]")
    print("   **Symptom:** [What the error looked like]")
    print("   **Cause:** [What was actually wrong]")
    print("   **Fix:** [What solved it]")
    print("")
    print("   The file needs paths: frontmatter so it auto-loads for relevant files.")
    print("   Keep entries to 3 lines. These load into context automatically via Claude rules.")


def main():
    # Read hook input from stdin — only proceed on actual git commit commands
    try:
        hook_input = json.loads(sys.stdin.read())
    except (json.JSONDecodeError, ValueError) as e:
        print(f"⚠️  on-commit: Could not parse hook input: {e}", file=sys.stderr)
        hook_input = {}

    tool_input = hook_input.get('tool_input', {})
    command = tool_input.get('command', '') if isinstance(tool_input, dict) else str(tool_input)
    if not re.search(r'\bgit\s+commit\b', command):
        sys.exit(0)

    # Check if the commit actually succeeded — failed commits have error output
    tool_response = hook_input.get('tool_response', '')
    response_str = str(tool_response).lower() if tool_response else ''
    if 'nothing to commit' in response_str or 'no changes added' in response_str:
        sys.exit(0)
    # Guard needed: 'nothing to commit' means git exited 0 but no commit was made — diff-tree HEAD would return unchanged files

    resolved = reset_loop_state()
    signal_learning_capture(resolved)
    sys.exit(0)


if __name__ == '__main__':
    main()

File: plugins/shipyard/bin/hooks/on-commit.mjs
/**
 * Post-commit hook: reset loop detection and trigger learning capture.
 *
 * Node port of project-files/scripts/on-commit.py.
 *
 * After a successful git commit:
 *  1. Reset the loop detection counter for committed files
 *  2. Detect if a struggle just resolved
 *  3. Signal the agent to capture learnings into .claude/rules/learnings/
 *
 * STDOUT CONTRACT: PostToolUse hook — stdout becomes a conversation
 * message to Claude. Learning prompts go to stdout intentionally.
 */

import { execFileSync } from "node:child_process";
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { atomicWrite, logEvent, resolveShipyardData, sanitizeForLog } from "../_hook_lib.mjs";

const DOMAIN_PATTERNS = [
  [["auth", "login", "session", "token", "jwt", "oauth"], "auth"],
  [["api", "route", "endpoint", "handler", "controller"], "api"],
  [["test", "spec", "__test", ".test.", ".spec."], "testing"],
  [["component", "page", "layout", "view", "ui"], "ui"],
  [["style", "css", "scss", "tailwind", "theme"], "styling"],
  [["model", "schema", "migration", "database", "db", "query", "supabase", "prisma"], "data"],
  [["hook", "context", "provider", "store", "state", "redux", "zustand"], "state"],
  [["config", "env", ".config", "setting"], "config"],
  [["util", "helper", "lib", "service", "action"], "logic"],
];

function detectDomain(filePath) {
  const lower = filePath.toLowerCase();
  for (const [keywords, domain] of DOMAIN_PATTERNS) {
    if (keywords.some((kw) => lower.includes(kw))) return domain;
  }
  return "general";
}

function getCommittedFiles() {
  try {
    const result = execFileSync(
      "git",
      ["diff-tree", "--no-commit-id", "--name-only", "-r", "HEAD"],
      { encoding: "utf8", stdio: ["ignore", "pipe", "pipe"], timeout: 5000 },
    );
    return new Set(result.trim().split("\n").map((s) => s.trim()).filter(Boolean));
  } catch (err) {
    if (err.code === "ENOENT") {
      process.stderr.write("⚠️  on-commit: git not found on PATH\n");
    } else {
      process.stderr.write(`⚠️  on-commit: git diff-tree failed: ${err.message}\n`);
    }
    return new Set();
  }
}

// Match task files: any .md file under a `spec/tasks/` directory.
export function isTaskFile(filepath) {
  return /(?:^|[/\\])spec[/\\]tasks[/\\][^/\\]+\.md$/i.test(filepath);
}

// Minimal YAML frontmatter parser — scalar string fields only. Mirrors the
// helper in tdd-check.mjs; they share the same minimal shape. Do not factor
// out prematurely — if either hook grows list/nesting needs, the two will
// diverge naturally.
export function parseTaskFrontmatter(content) {
  const m = content.match(/^---\s*\r?\n([\s\S]*?)\r?\n---/);
  if (!m) return {};
  const out = {};
  for (const line of m[1].split(/\r?\n/)) {
    const kv = line.match(/^([A-Za-z_][A-Za-z0-9_-]*)\s*:\s*(.*)$/);
    if (!kv) continue;
    let val = kv[2].trim();
    if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'"))) {
      val = val.slice(1, -1);
    }
    if (val.startsWith("#")) val = "";
    out[kv[1]] = val;
  }
  return out;
}

/**
 * For every committed task file whose frontmatter shows a `kind: operational`
 * task transitioning to `status: done` with a populated `verify_output`, emit
 * an `operational_task_passed` diagnostic event. Missing `operational_task_passed`
 * events for operational tasks that ARE marked done is the smoking-gun signal
 * for silent-pass regressions — the post-subagent gate in ship-execute is
 * supposed to emit it, and if it didn't but the task still landed as done, the
 * gate drifted. Emitting from this hook means the event fires regardless of
 * which path marked the task done.
 */
export function emitOperationalTaskPassedEvents(committedFiles, shipyardData) {
  for (const f of committedFiles) {
    if (!isTaskFile(f)) continue;
    let content;
    try {
      content = readFileSync(f, "utf8");
    } catch (_err) {
      continue;
    }
    const fm = parseTaskFrontmatter(content);
    if (fm.kind !== "operational") continue;
    if (fm.status !== "done") continue;
    const verifyOutput = fm.verify_output || "";
    if (!verifyOutput) continue; // tdd-check should have blocked this case
    logEvent(shipyardData, "operational_task_passed", {
      task_file: f,
      capture: verifyOutput,
    });
  }
}

function resetLoopState(loopStatePath) {
  if (!existsSync(loopStatePath)) return [];

  const committedFiles = getCommittedFiles();
  if (committedFiles.size === 0) return [];

  let state;
  try {
    state = JSON.parse(readFileSync(loopStatePath, "utf8"));
  } catch (err) {
    process.stderr.write(`⚠️  on-commit: Corrupt loop state, skipping reset: ${err.message}\n`);
    return [];
  }
  if (!state || typeof state !== "object") return [];
  if (!state.edits) state.edits = {};
  if (!state.struggles) state.struggles = {};

  const resolved = [];
  for (const filePath of committedFiles) {
    if (state.struggles[filePath]) {
      resolved.push({
        file: filePath,
        edit_count: state.struggles[filePath].edit_count || 0,
      });
    }
  }

  for (const f of committedFiles) {
    delete state.edits[f];
    delete state.struggles[f];
  }

  try {
    atomicWrite(loopStatePath, JSON.stringify(state, null, 2));
  } catch (err) {
    process.stderr.write(`⚠️  on-commit: Failed to save loop state: ${err.message}\n`);
  }

  return resolved;
}

function signalLearningCapture(resolved) {
  if (resolved.length === 0) return;

  const files = resolved.map((s) => s.file);
  const maxEdits = Math.max(...resolved.map((s) => s.edit_count));
  const domains = new Set(files.map(detectDomain));
  const domainHint = [...domains].sort().join(", ");
  const safeFiles = files.map((f) => sanitizeForLog(f, 200));

  process.stdout.write("\n");
  process.stdout.write("📝 LEARNING OPPORTUNITY — You just resolved a struggle.\n");
  process.stdout.write(`   Files: ${safeFiles.join(", ")}\n`);
  process.stdout.write(`   Edits before resolution: ${maxEdits}\n`);
  process.stdout.write(`   Suggested domain(s): ${domainHint}\n`);
  process.stdout.write("\n");
  process.stdout.write("   Append what you learned to .claude/rules/learnings/<domain>.md\n");
  process.stdout.write("   (create the file if it doesn't exist)\n");
  process.stdout.write("\n");
  process.stdout.write("   Format each entry as:\n");
  process.stdout.write("   ### [Short title]\n");
  process.stdout.write("   **Symptom:** [What the error looked like]\n");
  process.stdout.write("   **Cause:** [What was actually wrong]\n");
  process.stdout.write("   **Fix:** [What solved it]\n");
  process.stdout.write("\n");
  process.stdout.write("   The file needs paths: frontmatter so it auto-loads for relevant files.\n");
  process.stdout.write("   Keep entries to 3 lines. These load into context automatically via Claude rules.\n");
}

export async function run(hookInput, _env) {
  const shipyardData = (await resolveShipyardData()) || ".shipyard";
  const loopStatePath = join(shipyardData, ".loop-state.json");

  const toolInput = hookInput?.tool_input;
  let command = "";
  if (toolInput && typeof toolInput === "object") {
    command = toolInput.command || "";
  } else if (toolInput !== undefined && toolInput !== null) {
    command = String(toolInput);
  }

  if (!/\bgit\s+commit\b/.test(command)) return 0;

  const toolResponse = hookInput?.tool_response;
  const responseStr = toolResponse ? String(toolResponse).toLowerCase() : "";
  if (responseStr.includes("nothing to commit") || responseStr.includes("no changes added")) {
    return 0;
  }

  const resolved = resetLoopState(loopStatePath);
  signalLearningCapture(resolved);

  // Emit operational_task_passed events for any committed task files that
  // just transitioned an operational task to done. Early-return-fast if the
  // commit has no task files in it — keeps the hot path cheap on every
  // commit in the repo. Wrapped in try/catch because this is diagnostic-only
  // and must never fail the commit.
  try {
    const committedFiles = getCommittedFiles();
    emitOperationalTaskPassedEvents(committedFiles, shipyardData);
  } catch (err) {
    process.stderr.write(`⚠️  on-commit: operational_task_passed emission failed: ${err.message}\n`);
  }

  return 0;
}
